wallet/key_deriver.go

```diff
@@ -0,0 +1,153 @@
+	if counterparty.Type == CounterpartyTypeAnyone {
+		_, fixedKey := ec.PrivateKeyFromBytes([]byte{1})
+		counterparty = WalletCounterparty{
+			Type:         CounterpartyTypeOther,
+			Counterparty: fixedKey,
+		}
+	}
+
+	// Derive both public and private keys
+	derivedPublicKey, err := kd.DerivePublicKey(protocol, keyID, counterparty, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive public key: %w", err)
+	}
+
+	derivedPrivateKey, err := kd.DerivePrivateKey(protocol, keyID, counterparty)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive private key: %w", err)
+	}
+
+	// Create shared secret
+	sharedSecret, err := derivedPrivateKey.DeriveSharedSecret(derivedPublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create shared secret: %w", err)
+	}
+	if sharedSecret == nil {
+		return nil, fmt.Errorf("failed to derive shared secret")
+	}
+
+	// Return the x coordinate of the shared secret point
+	return sharedSecret.X.Bytes(), nil
+}
+
+func (kd *KeyDeriver) DerivePublicKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty, forSelf bool) (*ec.PublicKey, error) {
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute invoice number: %w", err)
+	}
+
+	var pubKey *ec.PublicKey
+	if forSelf {
+		privKey, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child private key: %w", err)
+		}
+		pubKey = privKey.PubKey()
+	} else {
+		pubKey, err = counterpartyKey.DeriveChild(kd.privateKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child public key: %w", err)
+		}
+	}
+	return pubKey, nil
+}
+
+func (kd *KeyDeriver) DerivePrivateKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty) (*ec.PrivateKey, error) {
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	k, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive child key: %w", err)
+	}
+	return k, nil
+}
+
+func (kd *KeyDeriver) NormalizeCounterparty(counterparty WalletCounterparty) *ec.PublicKey {
+	switch counterparty.Type {
+	case CounterpartyTypeSelf:
+		return kd.privateKey.PubKey()
+	case CounterpartyTypeOther:
+		return counterparty.Counterparty
+	case CounterpartyTypeAnyone:
+		_, pub := ec.PrivateKeyFromBytes([]byte{1})
+		return pub
+	default:
+		return nil
+	}
+}
+
+func (kd *KeyDeriver) ComputeInvoiceNumber(protocol WalletProtocol, keyID string) (string, error) {
+	// Validate protocol security level
+	if protocol.SecurityLevel < 0 || protocol.SecurityLevel > 2 {
+		return "", fmt.Errorf("protocol security level must be 0, 1, or 2")
+	}
```

Comment:
maybe it's worth to extaract a private function for validating protocol security level?

this would make the comment above redundant.

Reply:
We only care about matching their public interfaces, and only when we can

Reply:
Many cases where they TypeScript approach doesn't really map very directly to go